Mathlib/Tactic/Choose.lean

```diff
@@ -0,0 +1,240 @@
+/-
+Copyright (c) 2018 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl, Floris van Doorn, Mario Carneiro, Reid Barton, Johan Commelin
+-/
+
+import Mathlib.Util.Asserts
+import Mathlib.Util.Tactic
+import Mathlib.Tactic.ShowTerm
+
+import Mathlib.Lean.getBinderName
+import Mathlib.Lean.mkFreshNameFrom
+import Mathlib.Lean.addLocalVarInfoForBinderIdent
+
+open Lean Lean.Meta Elab.Term Elab.Tactic
+
+/-!
+# `choose` tactic
+Performs Skolemization, that is, given `h : ∀ a:α, ∃ b:β, p a b |- G` produces
+`f : α → β, hf: ∀ a, p a (f a) |- G`.
+
+TODO: switch to `rcases` syntax: `choose ⟨i, j, h₁ -⟩ := expr`.
+-/
+namespace function
+
+open Classical in
+/-- `sometimes f` evaluates to some value of `f`, if it exists. This function is especially
+interesting in the case where `α` is a proposition, in which case `f` is necessarily a
+constant function, so that `sometimes f = f a` for all `a`. -/
+noncomputable def sometimes {α β} [Nonempty β] (f : α → β) : β :=
+if h : Nonempty α then f (choice h) else choice ‹_›
+
+theorem sometimes_eq {p : Prop} {α} [Nonempty α] (f : p → α) (a : p) : sometimes f = f a :=
+dif_pos ⟨a⟩
+
+theorem sometimes_spec {p : Prop} {α} [Nonempty α]
+  (P : α → Prop) (f : p → α) (a : p) (h : P (f a)) : P (sometimes f) :=
+(sometimes_eq f a).symm ▸ h
+
+end function
+
+namespace Tactic
+
+namespace Choose
+
+/-- Given `α : Sort u`, `nonemp : Nonempty α`, `p : α → Prop`, a context of free variables
+`ctxt`, and a pair of an element `val : α` and `spec : p val`,
+`mk_sometimes u α nonemp p ctx (val, spec)` produces another pair `val', spec'`
+such that `val'` does not have any free variables from elements of `ctxt` whose types are
+propositions. This is done by applying `function.sometimes` to abstract over all the propositional
+arguments. -/
+def mk_sometimes (u : Level) (α nonemp p : Expr) :
+  List Expr → Expr × Expr → MetaM (Expr × Expr)
+| [], (val, spec) => pure (val, spec)
+| (e :: ctxt), (val, spec) => do
```

Comment:
Minor, but "context" is universally elsewhere abbreviated `ctx`, rather than `ctxt`. (It's also used inconsistently in the doc-string.)

Reply:
Fixed